### ISSUE TITLE
Add rule S6304: Policies granting access to all resources of an accounts are security-sensitive

### ISSUE THREAD

**Issue by pierre-loup-tristant-sonarsource** [SonarSource/SonarJS#3451]
# What 

The goal of this rule is to detect Python code using the AWS CDK that creates policies granting access to all resources. 
This rule has been previously implemented for Python. See [SONARPY-1162](https://sonarsource.atlassian.net/browse/SONARPY-1162).

# Detection logic

## `PolicyStatement`

Policies are defined by classes such as `Policy` or `ManagedPolicy`. These classes accept a `PolicyDocument`, which contains multiple `PolicyStatement`s. This `PolicyStatement` class should be the main focus of this rule.

### Constructor

The [constructor](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam.PolicyStatement.html) receives a single `PolicyStatementProps` parameters. The following checks should be applied to the parameters of the `PolicyStatementProps` object that the constructor receives:

Status | `resources` | `effect`
-- | -- | --
Noncompliant | <ul><li> `["*"]` (single element) </li><li>`["arn:aws:service:::a_resource/*","*"]` (contains `*` amount other elements) | <ul><li>`iam.Effect.ALLOW`</li><li>Not set
Compliant | Anything other than <ul><li> `["*"]` (single element) </li><li>`["arn:aws:service:::a_resource/*","*"]` (contains `*` amount other elements) | Any

**Exception**

If all actions defined in actions are KMS actions (i.e., all actions start with "kms:"), then usage of the wildcard identifier "*" in resources is compliant. See https://sonarsource.atlassian.net/browse/SONARIAC-260 for more details.

**Example**
```js
import { aws_iam as iam } from 'aws-cdk-lib'

const dummyPolicy = new iam.ManagedPolicy(
    // Dummy policy, removed for brevity
)

new iam.PolicyStatement({
    sid: "AllowAllResources",
    effect: iam.Effect.ALLOW, // secondary location
    actions: ["iam:CreatePolicyVersion"], 
    resources: ["*"] // Noncompliant
})

new iam.PolicyStatement({
    sid: "AllowSomeResources",
    effect: iam.Effect.ALLOW
    actions: ["iam:CreatePolicyVersion"], 
    resources: [dummyPolicy.managedPolicyArn] // Compliant
})

```

### `fromJson`

Apart from the constructor, the static method `PolicyStatement.fromJson()` may also be used to create a `PolicyStatement`. It takes an object as the first parameter, which should be checked as follows (using JSONPath syntax):

Status | `$.Resource` | `$.Effect`
-- | --| --
Noncompliant | <ul><li>`*`</li><li> `["*"]` (single element) </li><li>`["arn:aws:service:::a_resource/*","*"]` (contains `*` amount other elements) | <ul><li>`"Allow"` (case insensitive) </li><li>`iam.Effect.ALLOW`
Compliant | Anything other than <ul><li>`*`</li><li> `["*"]` (single element) </li><li>`["arn:aws:service:::a_resource/*","*"]` (contains `*` amount other elements)  | Any

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

const dummyPolicy = new iam.ManagedPolicy(
    // Dummy policy, removed for brevity
)

new iam.PolicyStatement.fromJson({
    "Sid": "AllowAllResources",
    "Effect": "Allow", // secondary location
    "Action": ["iam:CreatePolicyVersion"], 
    "Resource": ["*"] // Noncompliant
}

new iam.PolicyStatement.fromJson({
    "Sid": "AllowSomeResources",
    "Effect": "Allow",
    "Action": ["iam:CreatePolicyVersion"], 
    "Resource": [dummyPolicy.managedPolicyArn] // Compliant
})
```

### `fromJson`

`PolicyDocument` also has a static method `PolicyDocument.fromJson()` which will also indirectly create `PolicyStatements`. This method also takes a object, where `$.Statement` contains a `PolicyStatement[]` which is equal to the input of `PolicyStatement.fromJson()`.

The checks applied to `PolicyStatement.fromJson()` must be applied to each element of `$.Statement`.

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

const dummyPolicy = new iam.ManagedPolicy(
    // Dummy policy, removed for brevity
)

const policyDocument = {
    "Version": "2012-10-17",
    "Statement": [{
      "Sid": "AllowActionsList",
        "Effect": "Allow", // secondary location
        "Action": ["iam:CreatePolicyVersion"], 
        "Resource": ["*"] // Noncompliant
    },
    {
      "Sid": "AllowActionsStar",
        "Effect": "Allow", // secondary location
        "Action": ["iam:CreatePolicyVersion"], 
        "Resource": "*" // Noncompliant
    },
    {
        "Sid": "AllowSomeResources",
        "Effect": "Allow",
        "Action": ["iam:CreatePolicyVersion"], 
        "Resource": [dummyPolicy.managedPolicyArn] // Compliant
    }]
}

const policyDocumentFromJson = iam.PolicyDocument.fromJson(policyDocument)
```

## RSPEC

This [RSPEC PR](https://github.com/SonarSource/rspec/pull/1314) contains the message and highlighting guidelines.

**Comment by loris-s-sonarsource** [comment]
✅ LGTM!

**Comment by loris-s-sonarsource** [comment]
✅ Spec LGTM!
